docs/design/2022-07-28-support-generate-auto_random-pk.md

```diff
@@ -0,0 +1,100 @@
+    * [Use IGNORE_AUTO_RANDOM_COLUMN_VALUE to ignore the explicit primary key.](#use-ignore_auto_random_column_value-to-ignore-the-explicit-primary-key)
+  * [Compatibility](#compatibility)
+  * [Test Design](#test-design)
+    * [Functional Tests](#functional-tests)
+    * [Scenario Tests](#scenario-tests)
+
+## Introduction
+
+This design doc is about how to support allocating auto_random primary key in flink-tidb-connector.
+
+## Motivation or Background
+
+[AUTO_RANDOM](https://docs.pingcap.com/tidb/dev/auto-random) is a feature introduced by TiDB in order to increase the performance of the insert operation 
+when the writing hotspot issue occurs caused by auto-increment.
+
+The primary key, which must be bigint type and clustered, is allocated in the following way:
+The highest n(default is five，1 <= n < 16) digits (ignoring the sign bit) of the row value in binary (namely, shard bits) are determined by the starting time of the current transaction. 
+The remaining digits are allocated values in an auto-increment order.
+
+Luckily, we already support auto_increment and [shard_row_id_bits](https://docs.pingcap.com/tidb/dev/shard-row-id-bits) in Flink,
+of which the mechanism of the latter is similar to the auto_random, so we can reuse the methods.
+
+## Detailed Design
+
+Firstly, we will introduce the mechanism of shard and then discuss how wo get shardBits and get the shardSeed with good collision resistance.
+
+### the mechanism of shard
+
+Firstly, we need to generate the shardSeed, whose bits' length must be greater than and equal to 16.
+Then we take the low n bits of the shardSeed as the shard mask.
+
+```java
+long shardSeed = 0xffffL;
+long shardBits = 6L; 
+//((1L << shardBits) - 1) = 0b111111L ,shardMask = 0b101010L
+long shardMask = shardSeed & ((1L << shardBits) - 1);
+```
+
+In order to shard the highest n bits, we need to left shift the shardMask by (BigintLength - shardBits - signBitLength) bits.
+- The bigintLength is 64.
+- The signBitLength is 0 with unsigned bigint, and 1 with signed bigint.
+
+```text
+// example with unsigned bigint
+0b 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0010 1010   shardMask before left shift
+                                left shift                                          
+0b 1010 1000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000   shardMask after left shift
+                             bit or operation
+0b 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0001   auto_increment-generated-value
+                                 equals
+0b 1010 1000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0001   auto_random value
+```
+
+### Get the shardBits 
+
+When the [autoRandomBits](https://github.com/tidb-incubator/TiBigData/blob/master/tidb/src/main/java/io/tidb/bigdata/tidb/meta/TiTableInfo.java#L61) greater than 0, 
```

Comment:
do we need to judge if the shardBits>=16 and throw exception?

Reply:
TiDB will do the validation